src/audio_pixel.h

```diff
@@ -67,15 +104,55 @@ class audio_pixel_block_t {
 // and is able to interpolate between them to 
 // get audio pixels at any resolution inbetween
 class audio_pixel_mipmap_t {
-    audio_pixel_mipmap_t(MediaItem_Take* take);
-
-    void update(); // update all blocks
+    audio_pixel_mipmap_t(MediaItem_Take* take)
+        :m_accessor(safe_audio_accessor_t(take)),
+        m_take(take)
+    {
+        if (!m_accessor.is_valid()) { std::cerr << "Invalid audio accessor used for audio block." << std::endl; }
+    };
+
+    void update() { // update all blocks
+        // MUST BE CALLED BY MAIN THREAD
+        // due to reaper api calls: AudioAccessorUpdate
+        AudioAccessor* safe_accessor = m_accessor.get();
+
+        if (AudioAcessorStateChange(safe_accessor)) {
+            // updates the audio acessors contents
+            AudioAccessorUpdate(safe_accessor);
+
+            std::vector<double> sample_buffer;
+            double accessor_start_time = GetAudioAccessorStartTime(safe_accessor);
+            double accessor_end_time = GetAudioAccessorEndTime(safe_accessor);
+
+            // retrive the takes sample rate and num channels from PCM source
+            PCM_source* source = GetMediaItemTake_Source(m_take);
+            if (!source) { return; } // TODO: LOG ME
+            int sample_rate = GetMediaSourceSampleRate(source);
+            int num_channels = GetMediaSourceNumChannels(source);
+
+            // calculate the number of samples we want to collect per channel
+            // TODO CEILING ROUNDING
+            int samples_per_channel = sample_rate * (accessor_end_time - accessor_start_time);
+
+            // samples stored in sample_buffer, operation status is returned
+            int get_samples_status = GetAudioAcessorSamples(safe_accessor, sample_rate, num_channels,
+                accessor_start_time, samples_per_channel, sample_buffer.data());
+            if (get_samples_status != 1) { return; } // TODO: LOG ME
```

Comment:
i would rather have this return value called `errcode` or something like that

Reply:
if we're treating all failure cases the same, we don't even need to assign a variable to it 👍🏽 
```cpp

if (!GetAudioAccessorSamples(...)) 
   { return; }
```